### ISSUE TITLE
Phase 62.2: SpeechRecognizer — End-to-End Automatic Speech Recognition

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#1163]
## Phase 62.2: SpeechRecognizer — End-to-End Automatic Speech Recognition

> **Parent discussion**: #1161 (Phase 62 Planning — Speech & Audio Intelligence)
> **Depends on**: #1162 (Phase 62.1 AudioFeatureExtractor)

### Overview

The `SpeechRecognizer` implements end-to-end automatic speech recognition (ASR) supporting CTC decoding, attention-based sequence-to-sequence models, Conformer architecture, and Whisper-style multilingual recognition. Includes language model rescoring, beam search, and streaming recognition with partial hypotheses.

### Architecture

```mermaid
classDiagram
    class SpeechRecognizer {
        -config: ASRConfig
        -encoder: AudioEncoder
        -decoder: ASRDecoder
        -ctcModule: CTCDecoder
        -lmRescorer: LanguageModelRescorer
        -tokenizer: SpeechTokenizer
        +recognize(audio: AudioBuffer): TranscriptionResult
        +recognizeStreaming(chunk: AudioChunk): PartialTranscription
        +recognizeBatch(audios: AudioBuffer[]): TranscriptionResult[]
        +detectLanguage(audio: AudioBuffer): LanguageDetection
        +getTimestamps(audio: AudioBuffer): AlignedTranscription
    }

    class AudioEncoder {
        -conformerBlocks: ConformerBlock[]
        -subsamplingLayer: ConvSubsampling
        +encode(features: FeatureMatrix): EncoderOutput
        +encodeStreaming(chunk: FeatureMatrix, state: EncoderState): StreamingEncoderOutput
    }

    class ConformerBlock {
        -feedForward1: FeedForwardModule
        -selfAttention: MultiHeadSelfAttention
        -convModule: ConvolutionModule
        -feedForward2: FeedForwardModule
        +forward(x: Tensor, mask?: Tensor): Tensor
    }

    class CTCDecoder {
        -blankId: number
        -beamWidth: number
        +greedyDecode(logits: Float32Array[]): number[]
        +beamSearch(logits: Float32Array[], beamWidth: number): BeamHypothesis[]
        +prefixBeamSearch(logits: Float32Array[], lmWeight: number): BeamHypothesis[]
    }

    class ASRDecoder {
        -transformerDecoder: TransformerDecoder
        -outputProjection: LinearLayer
        +decode(encoderOutput: EncoderOutput, prevTokens: number[]): DecoderOutput
        +beamSearch(encoderOutput: EncoderOutput, beamSize: number): BeamHypothesis[]
        +jointCTCAttention(encoderOutput: EncoderOutput, ctcWeight: number): TranscriptionResult
    }

    class LanguageModelRescorer {
        -lmModel: LanguageModel
        -lmWeight: number
        +rescore(hypotheses: BeamHypothesis[]): BeamHypothesis[]
        +shallowFusion(beamState: BeamState, lmWeight: number): BeamState
    }

    class SpeechTokenizer {
        -vocabulary: Map~number, string~
        -bpeModel?: BPEModel
        +encode(text: string): number[]
        +decode(tokens: number[]): string
        +getVocabSize(): number
    }

    SpeechRecognizer --> AudioEncoder
    SpeechRecognizer --> ASRDecoder
    SpeechRecognizer --> CTCDecoder
    SpeechRecognizer --> LanguageModelRescorer
    SpeechRecognizer --> SpeechTokenizer
    AudioEncoder --> ConformerBlock
```

### Interface Definitions

```typescript
interface ASRConfig {
  modelType: "ctc" | "attention" | "hybrid" | "whisper";
  encoderLayers: number;          // Number of Conformer blocks (default: 12)
  encoderDim: number;             // Encoder hidden dimension (default: 256)
  decoderLayers: number;          // Decoder transformer layers (default: 6)
  decoderDim: number;             // Decoder hidden dimension (default: 256)
  numHeads: number;               // Attention heads (default: 4)
  convKernelSize: number;         // Conformer conv kernel (default: 31)
  vocabSize: number;              // Output vocabulary size
  ctcWeight: number;              // CTC loss weight in hybrid (default: 0.3)
  beamSize: number;               // Beam search width (default: 10)
  maxLength: number;              // Max output length
  language?: string;              // Target language code (ISO 639-1)
  streaming: StreamingConfig;
  lmConfig?: LMRescoringConfig;
}

interface StreamingConfig {
  enabled: boolean;
  chunkSizeMs: number;            // Chunk size in ms (default: 640)
  leftContext: number;            // Left context frames
  rightContext: number;           // Right (lookahead) context frames
  endpointingConfig: EndpointingConfig;
}

interface EndpointingConfig {
  silenceDurationMs: number;      // Silence duration to trigger endpoint (default: 700)
  maxUtteranceDurationMs: number; // Max utterance duration (default: 30000)
  trailingContextMs: number;      // Context after endpoint detection
}

interface TranscriptionResult {
  text: string;
  tokens: TokenInfo[];
  confidence: number;
  language?: string;
  duration: number;
  processingTime: number;
  alternatives?: AlternativeTranscription[];
}

interface PartialTranscription {
  stableText: string;             // Finalized text
  unstableText: string;           // Tentative text (may change)
  isFinal: boolean;
  confidence: number;
}

interface TokenInfo {
  token: string;
  startTime: number;
  endTime: number;
  confidence: number;
}

interface AlignedTranscription {
  words: WordAlignment[];
  text: string;
}

interface WordAlignment {
  word: string;
  startTime: number;
  endTime: number;
  confidence: number;
}

interface BeamHypothesis {
  tokens: number[];
  score: number;
  lmScore: number;
  ctcScore: number;
  attentionScore: number;
}

interface LanguageDetection {
  language: string;
  confidence: number;
  topK: Array<{ language: string; probability: number }>;
}
```

### Testing Strategy

1. **CTC decoding**: Verify greedy and beam search against known logit sequences
2. **Conformer forward pass**: Test output dimensions and attention mask handling
3. **Streaming parity**: Ensure streaming recognition matches batch within WER tolerance
4. **Beam search**: Verify beam pruning and LM rescoring improves hypotheses
5. **Language detection**: Test multilingual model language identification accuracy
6. **Timestamp alignment**: Verify word-level timestamps against forced alignment reference
7. **Edge cases**: Empty audio, noise-only, very short utterances, code-switching

### References

- Graves et al. (2006) — *Connectionist Temporal Classification* (ICML)
- Gulati et al. (2020) — *Conformer: Convolution-augmented Transformer for Speech Recognition* (Interspeech)
- Radford et al. (2023) — *Robust Speech Recognition via Large-Scale Weak Supervision* (Whisper, ICML)
- Chan et al. (2016) — *Listen, Attend and Spell* (LAS, ICASSP)
- Kim et al. (2017) — *Joint CTC-Attention Based End-to-End Speech Recognition* (IEEE/ACM TASLP)
- Graves (2012) — *Sequence Transduction with Recurrent Neural Networks* (RNN-T)
